server/routes/support/handlers/offenderLicenceDates.ts

```diff
@@ -0,0 +1,80 @@
+import { Licence } from '../../../@types/licenceApiClientTypes'
+import SimpleDate from '../../creatingLicences/types/date'
+import LicenceDatesAndReason from '../types/licenceDatesAndReason'
+
+export default class OffenderLicenceDatesRoutes {
+  constructor(private licenceService: LicenceService, private licenceOverrideService: LicenceOverrideService) {}
+
+  GET = async (req: Request, res: Response): Promise<void> => {
+    const { user } = res.locals
+    const { licenceId } = req.params
+
+    const licence = await this.licenceService.getLicence(licenceId, user)
+    const licenceDates = await this.getLicenceSimpleDates(licence)
+    res.render('pages/support/offenderLicenceDates', {
+      licence,
+      licenceDates,
+    })
+  }
+
+  POST = async (req: Request, res: Response): Promise<void> => {
+    const { user } = res.locals
+    const { licenceId, nomsId } = req.params
+    const { dateChangeReason } = req.body
+    const updatedDates = await this.getDatesFromSimpleDates(req.body)
+
+    if (updatedDates && dateChangeReason) {
+      await this.licenceOverrideService.overrideDates(parseInt(licenceId, 10), updatedDates, dateChangeReason, user)
+      res.redirect(`/support/offender/${nomsId}/licences`)
+      return
+    }
+
+    const licence = await this.licenceService.getLicence(licenceId, user)
+    const licenceDates = await this.getLicenceSimpleDates(licence)
+    res.render('pages/support/offenderLicenceDates', {
+      licence,
+      licenceDates,
+      dateChangeReason,
+    })
+  }
+
+  getLicenceSimpleDates = async (licence: Licence): Promise<object> => {
+    return {
+      crd: dateStringToSimpleDate(licence.conditionalReleaseDate),
+      ard: dateStringToSimpleDate(licence.actualReleaseDate),
+      ssd: dateStringToSimpleDate(licence.sentenceStartDate),
+      sed: dateStringToSimpleDate(licence.sentenceEndDate),
+      lsd: dateStringToSimpleDate(licence.licenceStartDate),
+      led: dateStringToSimpleDate(licence.licenceExpiryDate),
+      tussd: dateStringToSimpleDate(licence.topupSupervisionStartDate),
+      tused: dateStringToSimpleDate(licence.topupSupervisionExpiryDate),
+    }
+  }
+
+  getDatesFromSimpleDates = async (formData: LicenceDatesAndReason): Promise<object> => {
+    return {
+      conditionalReleaseDate: this.simpleDateToLicenceDate(formData.crd),
+      actualReleaseDate: this.simpleDateToLicenceDate(formData.ard),
+      sentenceStartDate: this.simpleDateToLicenceDate(formData.ssd),
+      sentenceEndDate: this.simpleDateToLicenceDate(formData.sed),
+      licenceStartDate: this.simpleDateToLicenceDate(formData.lsd),
+      licenceExpiryDate: this.simpleDateToLicenceDate(formData.led),
+      topupSupervisionStartDate: this.simpleDateToLicenceDate(formData.tussd),
+      topupSupervisionExpiryDate: this.simpleDateToLicenceDate(formData.tused),
+    }
+  }
+
+  simpleDateToLicenceDate = (date: SimpleDate): string | undefined => {
+    if (!date) {
+      return undefined
```

Comment:
What happens if you only enter a new value for one date (eg CRD) and leave the rest blank? Will this not override the values stored in the DB to `undefined` (or `null`)?

Reply:
Nevermind, just seen that this is handled in the njk file with `formResponses.<date> or licenceDates.<date>`